api/main.py

```diff
@@ -1204,5 +1206,115 @@ async def get_ci_badge_get(repo: str, branch: str, workflow:str):
     return Response(content=str(badge), media_type="image/svg+xml")
 
 
+class EnergyData(BaseModel):
+    type: str
+    company: Optional[UUID] = None
+    machine: UUID
+    project: Optional[UUID] = None
+    tags: Optional[str] = None
+    time_stamp: str
+    energy_value: str
+
+    @validator('company', 'project', 'tags', pre=True, always=True)
+    def empty_str_to_none(self, v):
+        if v == '':
+            return None
+        return v
+
+@app.post('/v1/carbondb/add')
+async def carbondb_add(request: Request, energydatas: List[EnergyData]):
+
+    client_ip = request.headers.get("x-forwarded-for")
+    if client_ip:
+        client_ip = client_ip.split(",")[0]
+    else:
+        client_ip = request.client.host
+
+    latitude, longitude = get_geo(client_ip)
+    carbon_intensity = get_carbon_intensity(latitude, longitude)
+
+    query_list = []
+    params_list = []
+    for e in energydatas:
+
+        e = html_escape_multi(e)
+
+        fields_to_check = {
+            'type': e.type,
+            'energy_value': e.energy_value,
+            'time_stamp': e.time_stamp,
+        }
+
+        for field_name, field_value in fields_to_check.items():
+            if field_value is None or field_value.strip() == '':
+                raise RequestValidationError(f"{field_name.capitalize()} is empty. Ignoring everything!")
+
+        if DB().fetch_one('SELECT id FROM carbondb_energy_data WHERE time_stamp=%s AND machine=%s AND energy_value=%s',
```

Comment:
Why does this guard clause exist? Is that a typical problem?

In larger data ingests this seems rather inefficient. 

Maybe just do batch insert and not line by line and a de-duplication after all inserts?

Reply:
Changed to batch insert